### DISCUSSION TITLE
gpt3 commit automatic summarization

### DISCUSSION THREAD

**Discussion by NotBrianZach** [magit/magit#4825] [Ideas]
A cool feature would be to query a language model api to automatically generate commit messages 
https://github.com/KanHarI/gpt-commit-summarizer
e.g. this (but I would prefer to have it in my magit commit buffer rather than as a github action)

**Comment by NotBrianZach** [comment]
also I tried to get gpt3 to write this feature for me but i dont think it totally works 

```emacs-lisp
(defun call-gpt3-api ()
  "Call GPT-3 API to get suggested commit message"
  (let ((url "https://api.openai.com/v1/engines/davinci/completions")
  (headers (list (cons "Content-Type" "application/json")
  (cons "Authorization" (format "Bearer %s" (openai-access-token)))))
  (params `(("prompt" . ,(magit-get-commit-prompt))
  ("max_tokens" . 50)
  ("temperature" . 0.7)
  ("frequency_penalty" . 0.2)
  ("stop" . "[stop]")
  ("n" . 1))))
  (request url
  :type "POST"
  :data (json-encode params)
  :headers headers
  :parser 'json-read
  :success (cl-function
  (lambda (&key data &allow-other-keys)
  (let ((completions (cdr (assoc 'choices (assoc 'choices data))))
   (if completions
   (car (car completions))
   (magit-get-commit-prompt))))))))mit message inside magit using the gpt3 api

(defun magit-auto-suggest-commit-message ()
  "Automatically suggest commit message using GPT-3 API"
  (interactive)
  (let ((message (call-gpt3-api)))
  (magit-set-commit-message message)))
```

**Reply by tarsius** [reply]
Not nearly enough monkeys and typewriters were uses.

**Comment by tarsius** [comment]
Good commit messages do not only describe *what* but also *how* and *why*. Since the AI does not understand those things, it can at "best" produce something that sounds like it does. If it succeeds in convincing the reviewer that it does understand, then that is actually the *worst* outcome. I prefer a *obviously* lazy, incomplete commit message over a wrong commit message. At least those are instantly recognizable.

By having an AI write the commit message, you are also robbing yourself of an opportunity to once more think the change through. At this point your mind is in a different state than when writing the code, and you might spot issues, that previously eluded you.

Even if that wasn't the case, and I though this was a good idea, this would not have to be implemented in magit/git-commit. There are hooks that can be used for this sort of thing. But I would recommend using a plain old algorithm to write a commit message *stub* instead of using an AI to produce nonsense.

If you use an AI, I would request that you append a note to that effect to each generated commit message. If you use an AI to write a message for a commit in one of my repositories, then I expect you to declare that upfront and many other maintainers will do the same.

Also see stackoverflow's AI policy at https://stackoverflow.com/help/gpt-policy.

**Reply by NotBrianZach** [reply]
>  There are hooks that can be used for this sort of thing. But I would recommend using a plain old algorithm to write a commit message stub instead of using an AI to produce nonsense.

are the hooks you mention in magit itself, e.g. can i define a hook in my emacs config that will replace the default magit commit message? (i'm sure it's possible I just didnt see an immeadiately applicable function in helm or in the git-commit.el file in magit)

**Reply by tarsius** [reply]
Use `git-commit-setup-hook`.
